Mesh documentation improvement

artokin opened the pull request:
This task is not yet completed, please do not merge until all content is available.
More commits will be added later on.

This is attempt to solve issues found from https://confluence.arm.com/display/IoTBU/Mesh+-+Mbed+OS+documentation+hackathon+2019-04-16 that was arranged because of https://jira.arm.com/browse/ISGDEVPREQ-1311 and corresponding epic.

@AnotherButler , @mikter , @SeppoTakalo

Commit:
Document update

Added documentation for RF driver extensions API.
Added documentation about SW MAC IEEE 802.15.4-2015 subset feature extension support.

Commit:
6LoWPAN-ND documentation cleanup

-Remove Nanostack architecture from 6LoWPAN-ND introduction
-Remove removed chapters from docs.json
-Rename quick_start_intro.md to 6lowpan_nd_intro.md
-Add 6LoWPAN-ND specific content to 6LoWPAN intro.
-Rename header from "mesh-tech" to "6LoWPAN-ND-tech" and update links.
-Use 6LoWPAN-ND instead of 6LoWPAN in many cases to avoid confusion

Commit:
Update Nanostack binary sizes

-Add Wi-SUN BR and router binary size
-Update binary sizes when using ARM Compiler 6.11

Commit:
6LoWPAN Mesh - common name for Nanostack mesh

-Use term "6LoWPAN Mesh" instead of IP-based Mesh
-Update 6LoWPAN Mesh section in connectivity

Commit:
New documents: mesh_tutorial and nanostack_intro

Added new documents (initial drafts):
 - mesh_tutorial.md added to tutorials
 - nanostack_intro.md locates in references/technology/mesh but it is
   not yet referenced

Commit:
Add Wi-SUN to introduction

Commit:
Update mesh api section.

Commit:
Minor update to links

Commit:
Moved Picture to different folder.

Commit:
Update Mesh configuration docs

Commit:
Update mesh tutorial and configuration

artokin commented:
@AnotherButler , @SeppoTakalo , @mikter would you please review. 
The major changes we had in mind are now made and your review/contribution/correction is now needed.

Commit:
Edit MeshInterface.md

Make minor copy edits to first file.

AnotherButler commented on docs/tutorials/using_apis/connectivity/mesh_tutorial.md:
```diff
@@ -0,0 +1,136 @@
+<h1 id="mesh-tutorial">Mesh tutorial</h1>
```
I like parts of this tutorial a lot, but it's lacking some cohesion. What is the main thing this tutorial is trying to accomplish? What does it teach you how to do?

artokin replied:
This tutorial should answer to question "What mesh protocol to select". It looks like we should include also https://jira.arm.com/browse/IOTTHD-3368 here as it might provide the missing glue.

AnotherButler commented on docs/tutorials/using_apis/connectivity/mesh_tutorial.md:
```diff
@@ -0,0 +1,136 @@
+
+## Introduction
+
+Mbed OS supports variety of 802.15.4 based Mesh protocols. There is support for [Wi-SUN](../reference/wisun-tech.html), [Thread](../reference/thread-tech.html) and [6LoWPAN-ND](../reference/6LoWPAN-ND-tech.html) protocols and they all have their own unique characteristics. For example, Thread is designed for Home automation use while Wi-SUN and 6LoWPAN-ND are targeted more for Utilities and Smart ...
```
I love this section and how it talks about what the different protocols are designed for. Could you please expand on when to use which protocol?

artokin replied:
Will be added soon :-)

artokin replied:
Updated in commit "Update mesh introduction"

AnotherButler commented on docs/tutorials/using_apis/connectivity/mesh_tutorial.md:
```diff
@@ -0,0 +1,136 @@
+Mbed OS supports variety of 802.15.4 based Mesh protocols. There is support for [Wi-SUN](../reference/wisun-tech.html), [Thread](../reference/thread-tech.html) and [6LoWPAN-ND](../reference/6LoWPAN-ND-tech.html) protocols and they all have their own unique characteristics. For example, Thread is designed for Home automation use while Wi-SUN and 6LoWPAN-ND are targeted more for Utilities and Smart ...
+
+
+## Mesh application design principles
```
How does this section relate to what this tutorial is trying to accomplish?

artokin replied:
We didn't find any better place for "Mesh App Design guidelines". So far they are not documented anywhere and we feel that they could be part of Mesh tutorial.  Would you have any suggestions for the better location?

AnotherButler commented on docs/porting/connectivity/MeshMAC.md:
```diff
@@ -183,6 +206,161 @@ Member|Description
 `parent_id` | Service user ID used to indentify the MAC service user. Optional.
 `phyMTU` | Maximum Transmission Unit (MTU) used by MAC. Standard 802.15.4 MAC must set 127.
 
+## SW MAC Information element's and Enhanced ACK support
```
Is this only for people porting Mbed OS?
This file seems to deviate from our porting guide template.

juhhei01 replied:
No It is actually for MAC-2015 version features users like wi-sun stack. NEW MAC actually need's driver porting instaruction but those are documented in this PR alread. @JarkkoPaso will update later that porting part anyway.

AnotherButler replied:
This file is only for porting, so should we move this somewhere else?

SeppoTakalo replied:
I believe that this is in correct locations.
This file is about porting MAC layer driver.
But tittle should change to make that obvious.

juhhei01 replied:
@SeppoTakalo Ok I will update Tittle. is this OK "IEEE 802.15.4 MAC sublayer extended APIs"

artokin replied:
Title changed in commit "Review corrections"

AnotherButler commented on docs/porting/connectivity/MeshInterface.md:
```diff
@@ -176,6 +176,87 @@ State|Description
 
 ![tx](https://s3-us-west-2.amazonaws.com/mbed-os-docs-images/tx_process.png)
 
+## MESH PHY extended driver support
```
Is this only for people porting Mbed OS?
This file seems to deviate from our porting guide template.

juhhei01 replied:
OK I got your point it definitely must be move to Porting guide. @artokin , @mikter and @JarkkoPaso.

juhhei01 replied:
@JarkkoPaso and @artokin We must add Jarkko's PR commit to this when it is merged our feature branch. Jarkko have re writed porting guide so This comment not need any action now.

artokin replied:
Commit "RF driver porting guide refactored" added to this PR

AnotherButler commented on docs/api/networkinterfaces/MeshInterface.md:
```diff
@@ -2,52 +2,29 @@
 
 [![View code](https://www.mbed.com/embed/?type=library)](https://os.mbed.com/docs/development/mbed-os-api-doxy/class_mesh_interface.html)
 
-## Mesh example
```
Is there a brief, clever example we can transclude here?

artokin replied:
The https://github.com/ARMmbed/mbed-os-example-mesh-minimal/blob/master/main.cpp is the only one we have and I'm afraid it is not too brief. However, the application is referenced in the "Related content" section.

SeppoTakalo replied:
Arto, we could include small 10 line snipped here, just to showcase that Mesh applications follow the same interface rules as any other interface.
Something like this:

# Mesh example

Following example connects to mesh network

```

#include "mbed.h"

int main(void)
{
	MeshInterface *mesh = MeshInterface::get_default_instance();

	int error = mesh->connect();
    if (error) {
        printf("Connection failed! error %d\n", error);
        return error;
    }

    printf("Connected!\n");

    UDPSocket sock;
    error = sock.open(mesh);
    if (error) {
    	printf("Failed to open socket, error %d\n", error);
    }

    // Now interface is connected, and I can communicate with Sockets
}
```

artokin replied:
Added in commit "Review corrections"

AnotherButler reviewed (commented):
Thanks for the PR 👍 

I've left some high-level questions and comments.

juhhei01 commented on docs/porting/connectivity/MeshMAC.md:
```diff
@@ -1,6 +1,9 @@
+- IEEE 802.15.4-2015 enhanced ack
+- Header and paylod Information elements set by MAC user. 
 
 ## SW MAC
```
@AnotherButler Is this SW MAC chapter then also at wrong file. Reason Why I put MAC updates there was that.

Commit:
Update mesh introduction

Add "find correct mesh protocol" section from mikter01.

Commit:
Remove nanostack_intro.md

-Remove nanostack_intro.md as partly obsoleted and hard to place to
 document structure.
-Fix some typos

SeppoTakalo commented on docs/porting/connectivity/MeshMAC.md:
```diff
@@ -1,6 +1,9 @@
 <h1 id="mac-port">MAC API porting</h1>
```
This tittle should change.
MAC API porting should actually refer that this IEEE 802.15.4 software or hardware MAC.

For example: `<h1 id="rf-mac-port">Porting IEEE 802.15.4 MAC driver</h1>`

juhhei01 replied:
OK

juhhei01 replied:
I have Fixed Tittle

SeppoTakalo commented on docs/porting/connectivity/MeshMAC.md:
```diff
@@ -183,6 +206,161 @@ Member|Description
+## SW MAC Information element's and Enhanced ACK support
+
+This chapter introduce how to use extented features.
+### Initialize API
```
All that content should already be in the Doxygen fields in the header.
Therefore you should drop ALL function descriptions from here, and only talk about the porting process, pitfalls, and link to the Doxygen for API reference.

juhhei01 replied:
@SeppoTakalo Removing those it will make this file quite a short and less information.

SeppoTakalo replied:
Short does not mean there is less information.
You should already have all this in the header file. If not, fix your header files first.

Doxygen should be good enough, so that this is not needed.

SeppoTakalo commented on docs/reference/technology/mesh/thread_intro.md:
```diff
@@ -130,11 +130,21 @@ Now, the Thread network is ready to accept new joiner devices.
 
 <span class="notes">**Note:** Thread uses hashing and elliptic curve algorithms for the secure communication. PSKd(s) and EUI64(s) are never transmitted in plain text over the peer to peer connection.</span>
 
+### Storing configuration settings to File system
```
I wonder if we should move rest of this file to a separate "tutorial" page.
It is not anymore talking about Thread in general, but it is more like a hands-on tutorial.

artokin replied:
Well, isn't that quite heavily with Thread. Storing configuration is built in to Thread, user just needs to activate it. Also commissioning is Thread related. 
I would like to keep the rest of the file as it is now.

SeppoTakalo reviewed (commented):
Some review remarks.

Commit:
Update MeshMAC.md

Tittle update

Commit:
RF driver porting guide refactored (#1071)

Commit:
Review corrections

-Add mesh example code snippet
-Update MeshMAC title

AnotherButler commented:
@SeppoTakalo Have all of the changes you requested been made?

SeppoTakalo commented:
Yes.
